text/0000-target-bundles.md

```diff
@@ -0,0 +1,245 @@
+JSON target specifications before packaging them for use by rustc. This way we get to keep using
+JSON and can keep on having comments in the checked-in target specifications.
+
+Similar preprocessing step could be used to implement some form of target inheritance so the
+duplication between built-in targets could be reduced greatly.
+
+[comments]: https://github.com/rust-lang/rust/pull/34980#issuecomment-234683183
+[jsmin]: https://plus.google.com/+DouglasCrockfordEsq/posts/RK8qyGVaGSr
+
+### Target quirks and cfg
+
+Currently we have a few keys dealing with the configuration variables only: `target_os`,
+`target_env`, `target_vendor`, `target_family`, `target_endian`, `target_pointer_width`¹, etc;
+Then there’s options from which some configuration variables are derived, but they are also used to
+tweak compilation: `is_like_windows`, `is_like_osx`, `is_like_solaris`, etc.
+
+This RFC proposes replacing these keys with a different set of keys which explicitly control
+configuration variables and keys which control compilation details:
+
+* `cfg: {"target_env": "msvc", windows: null }` would result in two cofiguration variables
+`#[cfg(windows)]` and `#[cfg(target_env = "msvc")]` evaluating to true, but no variables provided
+by the `cfg` key would get used to influence the behaviour of the compiler; and
+* `debuginfo: ["CodeView", 1]` some targets require different debuginfo format than what LLVM
+generates by default. MSVC targets want CodeView version 1, OS X and Android want Dwarf version 2,
+while LLVM appears to use the highest supported version of Dwarf by default.
+* similarly for many other variables which tweak the way compilation is done.
+
+¹: Technically, `target_pointer_width` is used in trans, but it does not provide any extra
+information over `data_layout`, which is what it should be using.
+
+This would allow selectively reusing various conditional implementations that are present in the
+compiler for a custom target, when none of the `is_like_*` variables would fully suit the target.
+Moreover, being able to specify arbitrary cfg variables would allow easily adapting for various
+miniscule details related to the targets. For example, the targets for ARM CPUs with NEON support
+could export `target_has_neon` without any extra language support.
+
+That being said, it might make sense to have a whitelist of stable options and cfg variables and
+keep everything else unstable for some duration.
+
+### Proposed JSON key-values
+
+```json
+{
+    // REQUIRED
+    "llvm-target": "x86_64-unknown-linux-gnu",              // LLVM target triple (does not need to match with rustc triple)
+    "data-layout": "e-m:e-i64:64-f80:128-n8:16:32:64-S128", // DataLayout for the target
+    "arch": "x86_64",                                       // Architecture of the target
+
+    // OPTIONAL
+    // Configuration variables injected into the compilation units.
+    "cfg": {
+        "target_os": "linux",
```

Comment:
I also meant `target_os`, which can't be derived from `data-layout`; it should have a comment about it being a mandatory field. I'm not sure about deriving `target_pointer_witdth` et al. from `data-layout` because we don't have control over its format/contents; it might stop containing information about the pointer width in the future.